pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -228,22 +257,37 @@ func (r *ReconcileMachineHealthCheck) Reconcile(request reconcile.Request) (reco
 }
 
 func isAllowedRemediation(mhc *mapiv1.MachineHealthCheck) bool {
+	maxUnhealthy, err := getMaxUnhealthy(mhc)
+	if err != nil {
+		return false
+	}
+
+	// If unhealthy is above maxUnhealthy, short circuit any further remediation
+	return unhealthyMachineCount(mhc) <= maxUnhealthy
+}
+
+func getMaxUnhealthy(mhc *mapiv1.MachineHealthCheck) (int, error) {
 	if mhc.Spec.MaxUnhealthy == nil {
-		return true
+		// This value should be defaulted, but if not, 100% is the default
+		return derefInt(mhc.Status.ExpectedMachines), nil
 	}
 	maxUnhealthy, err := getValueFromIntOrPercent(mhc.Spec.MaxUnhealthy, derefInt(mhc.Status.ExpectedMachines), false)
 	if err != nil {
 		glog.Errorf("%s: error decoding maxUnhealthy, remediation won't be allowed: %v", namespacedName(mhc), err)
-		return false
+		return 0, err
 	}
 
 	if maxUnhealthy < 0 {
 		maxUnhealthy = 0
 	}
 
-	// if noHealthy are above maxUnhealthy we short circuit any farther remediation
-	noHealthy := derefInt(mhc.Status.ExpectedMachines) - derefInt(mhc.Status.CurrentHealthy)
-	return (maxUnhealthy - noHealthy) >= 0
+	return maxUnhealthy, nil
+}
+
+// unhealthyMachineCount calculates the number of presently unhealthy or missing machines
+// ie the delta between the expected number of machines and the current number deemed healthy
+func unhealthyMachineCount(mhc *mapiv1.MachineHealthCheck) int {
+	return derefInt(mhc.Status.ExpectedMachines) - derefInt(mhc.Status.CurrentHealthy)
```

Comment:
I think it will default to 0 if the deref fails (I think that's the point of the function), but no matter the behaviour, this hasn't actually changed from before this PR, just moved

Reply:
ok perfect. thanks!